hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -3074,3 +3053,68 @@ abstract class NDArrayEmitter[C](
     columnMajorLoops
   }
 }
+
+abstract class NDArrayEmitter2(val outputShape: IEmitCodeGen[IndexedSeq[Value[Long]]])
+{
+  val nDims = outputShape.value.length
+
+  def outputElement(cb: EmitCodeBuilder, idxVars: IndexedSeq[Value[Long]]): Code[_]
+
+  def emit(cb: EmitCodeBuilder, targetType: PNDArray, region: Value[Region]): IEmitCode = {
+    outputShape.map(cb){ shapeArray: IndexedSeq[Value[Long]] =>
+      val dataSrvb = new StagedRegionValueBuilder(cb.emb, targetType.data.pType, region)
+      val dataAddress: Code[Long] =
+        Code(
+          dataSrvb.start(targetType.numElements(shapeArray, cb.emb).toI),
+          emitLoops(cb, shapeArray, targetType.elementType, dataSrvb),
+          dataSrvb.end())
+
+      def shapeBuilder(srvb: StagedRegionValueBuilder): Code[Unit] = {
```

Comment:
`IEmitCode` things shouldn't contain things of type `Code[Unit]` --- they should be functions of type `Unit` that add the instructions directly to the CodeBuilder.

Reply:
I'm not sure what you want me to change here. These `StagedRegionValueBuilder => Code[Unit]` functions are constructed this way to pass to `StagedRegionValueBuilder.addBaseStruct`, which expects a function of that type. Updating how SRVBs work seems potentially out of scope for this PR.

Reply:
Ah, I see. That comment was both for this and `emitLoops`, which I think can get changed to add code to `cb` directly. I agree, updating SRVB seems out of scope for this PR. But if I'm not mistaken, one of the things we were talking about doing was phasing out SRVB in favor of constructors on the PType, so having a non-SRVB-based constructor signature on PNDArray would be nice (not necessary for this PR).